### PR TITLE
 NUP-2506: Add test to all Serializable subclasses and fix related issues 

### LINKED ISSUE #3820: Saving and loading a model repeatedly causes it to break
Hi!

I'm feeding data to a model in small batches, saving the model to disk at the end of each batch, and loading it again for the next one. After a few batches, the model stops working and throws the following error when calling `model.run(input)`:
```
Traceback (most recent call last):
  File "./anomalies.py", line 63, in <module>
    result = model.run(input)
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/frameworks/opf/htm_prediction_model.py", line 448, in run
    inferences = self._anomalyCompute()
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/frameworks/opf/htm_prediction_model.py", line 696, in _anomalyCompute
    self._getAnomalyClassifier().compute()
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/engine/__init__.py", line 433, in compute
    return self._region.compute()
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/bindings/engine_internal.py", line 1499, in compute
    return _engine_internal.Region_compute(self)
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/bindings/regions/PyRegion.py", line 184, in guardedCompute
    return self.compute(inputs, DictReadOnlyWrapper(outputs))
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/regions/knn_anomaly_classifier_region.py", line 326, in compute
    self._classifyState(record)
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/regions/knn_anomaly_classifier_region.py", line 405, in _classifyState
    self._addRecordToKNN(state)
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/regions/knn_anomaly_classifier_region.py", line 490, in _addRecordToKNN
    knn.learn(pattern, category, rowID=rowID)
  File "/home/dani/.local/lib/python2.7/site-packages/nupic/algorithms/knn_classifier.py", line 537, in learn
    inputPattern = numpy.dot(self._vt, inputPattern - self._mean)
ValueError: operands could not be broadcast together with shapes (65536,) (0,)
```

Here's the code used to load and store the model:
```python
with open(model_file, 'r') as f:
    model = HTMPredictionModel.readFromFile(f)
```
```python
with open(model_file, 'w') as f:
    model.writeToFile(f)
```

I've tried using a model generated from a previous batch and skipping some batches of data, to find out if it was the data that was somehow generating a bad model, but after the same number of batches, no matter their contents, I get to a broken model again. Thus, I suspect a bug is being triggered at `readFromFile` or `writeToFile` (or maybe I'm just doing it wrong).

This is with Python 2.7.9, and nupic 1.0.3 from pypi.

### PR DESCRIPTION
**Pull request by lscheinkman**
@rhyolight Please review.
It should fix #3820, #3721, numenta/nupic.core#1065

> Note: This PR depends on https://github.com/numenta/nupic.core/pull/1417

### CODE REVIEW

File: src/nupic/regions/anomaly_region.py [hunk 1]
```diff
@@ -101,15 +107,20 @@ def __ne__(self, other):
...
     anomalyRegion = object.__new__(cls)
-    anomalyRegion.prevPredictedColumns = numpy.array(proto.prevPredictedColumns)
+    anomalyRegion.prevPredictedColumns = numpy.array(proto.prevPredictedColumns,
+                                                     dtype=numpy.float32)
     return anomalyRegion
 
 
   def write(self, proto):
-    proto.prevPredictedColumns = self.prevPredictedColumns.tolist()
+    proto.prevPredictedColumns = self.prevPredictedColumns.ravel().tolist()
```
Comment (rhyolight) [inline src/nupic/regions/anomaly_region.py]: 🤦‍♂️

File: src/nupic/algorithms/backtracking_tm.py [hunk 2]
```diff
@@ -470,8 +470,10 @@ def write(self, proto):
     proto.lrnIterationIdx = self.lrnIterationIdx
     proto.iterationIdx = self.iterationIdx
     proto.segID = self.segID
-    if self.currentOutput is not None:
-      proto.currentOutput = self.currentOutput.tolist()
+    if self.currentOutput is None:
+      proto.currentOutput.none = None
```
[the remark this answers is not in the document]
Comment (lscheinkman) [reply src/nupic/algorithms/backtracking_tm.py]: This is a way to handle 'None' in pycapnp. You first create a union with a 'Void' element representing 'None' and then initialize it with 'None'
see https://jparyani.github.io/pycapnp/quickstart.html#unions

File: src/nupic/encoders/delta.py [hunk 3]
```diff
@@ -118,9 +130,12 @@ def read(cls, proto):
     encoder._adaptiveScalarEnc = (
       AdaptiveScalarEncoder.read(proto.adaptiveScalarEnc)
     )
-    encoder._prevAbsolute = proto.prevAbsolute
-    encoder._prevDelta = proto.prevDelta
+    encoder._prevAbsolute = None if proto.prevAbsolute == 0 else proto.prevAbsolute
+    encoder._prevDelta = None if proto.prevDelta == 0 else proto.prevDelta
```
[the remark this answers is not in the document]
Comment (lscheinkman) [reply src/nupic/encoders/delta.py]: `pylint` did not complain about this style

**Comment by rhyolight** [comment]
👍 Merge it and I'll release.

**Pull request merged**